experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp

```diff
@@ -0,0 +1,136 @@
+//===- extract_kernel_pass.cpp - Pipeline the kernel -===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#include "alp/Transforms/Passes.h"
+#include "alp/Transforms/PassDetail.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+#include <unordered_map>
+
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/SCF/Transforms.h"
+#include "mlir/Dialect/SCF/Utils.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+
+using namespace mlir;
+
+namespace{
+  struct ModuloSchedulingPass : public ModuloSchedulingPassBase<ModuloSchedulingPass>
+  {
+    //ModuloScheduling(int unrollFactor):unrollFactor_(unrollFactor){}
+    ModuloSchedulingPass() = default;
+    ModuloSchedulingPass(const ModuloSchedulingPass &pass) {}
+    void getDependentDialects(DialectRegistry &registry) const override
+    {
+      registry.insert<scf::SCFDialect>();
+    }
+    void runOnFunction() override
+    {
+      // Get the current FuncOp operation being operated on.
+      auto f = getFunction();
+
+      scf::ForOp loop;
+
+      //  Unroll the kernel
+      f.walk([&](scf::ForOp forop)
```

Comment:
Nit: A run of clang-format on this would restructure things very differently.
I locally added this to my `.bashrc`:

```
function git-format-add-and-amend(){
  echo "git add $1 && git show --name-only | egrep \"*.(\.cpp|\.h)\" | xargs -i clang-format --style=file -i {}; git add $1; git commit --amend"
  git add $1 && git show --name-only | egrep "*.(\.cpp|\.h)" | xargs -i clang-format --style=file -i {}; git add $1; git commit --amend
}
```

Reply:
Thanks for this, I will copy as is!